src/raz_meta.rs

```diff
@@ -112,4 +119,67 @@ impl FirstLast for usize {
 	fn last() -> Self { usize::max_value() }
 }
 
-// TODO: HashMap
\ No newline at end of file
+/// Metadata for names in a raz tree.
+///
+/// Hash is implemented by a no-op, since the data here
+/// can be found elsewhere in the tree. This is not intended
+/// to be used outside of a raz.
+#[derive(Clone,Eq,PartialEq,Debug)]
+pub struct Names(pub HashMap<Name,()>);
+impl Hash for Names {
+	/// does nothing
+	fn hash<H:Hasher>(&self, _state: &mut H) {}
+}
+
+#[derive(Clone,Debug)]
+pub enum NameIndex {
```

Comment:
The problem that these solve is that when the navigation finds a location early with `SideChoice::Here`, the cursor algorithm needs a way to indicate that the final location is at the far end of a vector. I could have use a convention like "always split before the first element", but it would have further complicated the code and required explanations. Using the type system seemed a better option. But this means that every index type needs a way to indicate each end.